proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

```diff
@@ -0,0 +1,177 @@
+bailing. These placeholder types are as described:
+
+```swift
+// @class Foo turns into
+@available(*, unavailable, message: “This Objective-C class has only been forward declared; import its owning module to use it”)
+class Foo : NSObject {}
+
+// @protocol Bar turns into
+@available(*, unavailable, message: “This Objective-C protocol has only been forward declared; import its owning module to use it”)
+protocol Bar : NSObjectProtocol {}
+```
+
+Permitted usages of these types are intentionally limited. You will be able to use Objective-C and C declarations that refer to these types without issue.
+You will be able to pass around instances of these incomplete types from Swift to Objective-C and vice versa.
+You’ll also be able to call the set of methods provided by NSObject or NSObjectProtocol on instances.
+
+Using the type itself directly in Swift is forbidden by the attached unavailable attribute. This is to keep the impact of the change small and to prevent unsound declarations,
+such as declaring a new Swift class that inherits from or conforms to such a type. You will also not be able to create new instances of these types in Swift.
+
+To make these limitations more intuitive to users, a new diagnostic has been added. This diagnostic helps guide users when they attempt to access a member on the incomplete synthesized
+type, expecting the complete type. The new diagnostics are shown in the last 4 lines of this example:
+
+```swift
+swift-client-fwd-declared.swift:6:7: error: value of type 'Foo' has no member 'sayHello'
+myFoo.sayHello()
+~~~~~ ^~~~~~~~
+<unknown>:0: note: class 'Foo' is a placeholder for a forward declared Objective-C interface and may be missing members; import the definition to access the complete interface
+foo-bar-consumer.h:3:1: note: interface 'Foo' forward declared here
+@class Foo;
+^
+```
+
+The feature is gated behind a new frontend flag `-enable-import-objc-forward-declarations`. This flag is on by default for Swift version 6 and onwards. The flag is always disabled in the REPL, as
+allowing it leads to confusing behavior. The [existing ClangImporter diagnostics](https://forums.swift.org/t/pitch-improved-clangimporter-diagnostics/52687) will still
+properly warn users about forward declaration limitations, with or with out the feature enabled.
+
+## Source compatibility
+
+This change does have the potential to break source compatibility. This is simply a consequence
+of the ClangImporter importing new declarations that were previously hidden. That said,
+the impact should rare and easy to fix (just rename or qualify any ambiguous identifiers). 
+
+As mentioned, to address this, the feature is gated behind a flag for Swift 5 and before, and enabled
+by default in Swift 6.
+
+That said, the current implementation already passes the source compatibility suite and the source kit stress test.
```

Comment:
I think it's fair to say that the impact should be rare, but I would omit the part about it being "easy to fix", since it's much more subjective (a now-colliding decl could somehow be used widely throughout a person's code base, making the refactoring larger).

I would move the sentence about the source compatibility suite and SK stress test up here since. Perhaps something like this for the remainder of the section:

```suggestion
we believe a source compatibility break will be rare; our testing found that our
implementation passes the source compatibility suite and the SourceKit stress test.
Furthermore, the feature is gated behind a flag for Swift 5 and before, so breaking
changes would only occur by enabling the flag or using the new language mode.
```